### ISSUE TITLE
Serve single page application

### ISSUE THREAD

**Issue by ArtemAstakhov** [stablekernel/aqueduct#274]
Hi,
I'm trying to write a single page application with aqueduct on the server side. How can I send one specific .html file (which imports css and js) for any request?

I have tried these approaches https://aqueduct.io/docs/http/serving_files/, but that's not exactly what I need.

**Comment by joeconwaystk** [comment]
If I understand your question correctly, you want to navigate to `http://myapp.com` and it have the app serve the `index.html` file that contains your SPA? (If not, let me know.)

Name the file index.html and place it and your css and js into a single directory. This directory might be named something like `web/` and should be in the top-level directory of your Aqueduct application, i.e. at the same level where pubspec.yaml is.

Set up a router like so:

```dart
router.route("/*).pipe(new HTTPFileController("web/"));
```

This will serve `web/index.html` whenever you navigate to your server in the browser. When the browser makes requests for for the references JS and CSS files, it'll also pull those from `web`.

**Comment by joeconwaystk** [comment]
(Or do you want to deal with the pesky issue of someone entering an Angular2 client-side route into the browser and having it return you the index.html no matter what?)

**Comment by ArtemAstakhov** [comment]
Yes, my React client-side is responsible for page routing and I want to get index.html even if user enter `http://app.com/fsdgsdgsdfgdf`

**Comment by joeconwaystk** [comment]
We're figuring out the best way to do this (paging @erikist). Here's something that works for as much as I tested it, you can replace HTTPFileController in your routes with it. It redirects requests that accept `text/html` (and would otherwise yield a 404) back to "/". It'll include the original client route in the X-JS-Route header.

```dart
class ReroutingFileController extends HTTPFileController {
  ReroutingFileController(String directory) : super(directory);

  @override
  Future<RequestOrResponse> processRequest(Request req) async {
    Response potentialResponse = await super.processRequest(req);
    var acceptsHTML = req.innerRequest
        .headers.value(HttpHeaders.ACCEPT).contains("text/html");
    if (potentialResponse.statusCode == 404 && acceptsHTML)  {
        return new Response(302, {
          HttpHeaders.LOCATION: "/",
          "X-JS-Route": req.path.remainingPath
        }, null);
    }
    return potentialResponse;
  }
}
```

**Comment by ArtemAstakhov** [comment]
I'm not sure I changed your code correctly, but it works as I need. Thank you.

```
class ReroutingFileController extends HTTPFileController {
  ReroutingFileController(String directory) : super(directory);

  @override
  Future<RequestOrResponse> processRequest(Request req) async {
    var html = new File('web/index.html').openRead();
    Response potentialResponse = await super.processRequest(req);
    var acceptsHTML = req.innerRequest
        .headers.value(HttpHeaders.ACCEPT).contains("text/html");
    if (potentialResponse.statusCode == 404 && acceptsHTML)  {
      return new Response.ok(html)
        ..encodeBody = false
        ..contentType = new ContentType('text', 'html', charset: "utf-8");
    }
    return potentialResponse;
  }
}
```

**Comment by joeconwaystk** [comment]
Sure. There's two things that are gonna bite you here with those changes:

First, `openRead` opens a stream - if you return `potentialResponse`, then the stream never gets consumed and it'll leak. You can move the `openRead` into the `if` clause and you'll solve that problem.

Second, if you return the contents of the HTML file in this way, a multi-segment route will fail to fetch relative paths from the HTML. e.g., `GET /nonsense` will work correctly, but `GET /nonsense/other_stuff` will actually fail when the referenced CSS/JS use relative paths. (The browser will try and fetch `/nonsense/other_stuff/react.js` instead.)

**Comment by ArtemAstakhov** [comment]
Yeah, I met the second problem. Absolute paths work fine.

**Comment by joeconwaystk** [comment]
Closing this. Two conveniences are merged and will be available in 2.3.0. The snippet above becomes:

```dart
router.route("/*").pipe(new HTTPFileController("web/", onFileNotFound: (c, req) async {
  if (req.acceptsContentType(ContentType.HTML)) {
    /* maybe more logic */
    return new Response(302, ..., ...);
  }
  return new Response.notFound();
}));
```
